### PR TITLE
Type annotations for sqlalchemy.ext.horizontal_shard

### PR DESCRIPTION
**Pull request by ghost**
### Description
An attempt to annotate `lib/sqlalchemy/ext/horizontal_shard.py` with type hints (issue #6810).

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: lib/sqlalchemy/orm/session.py [hunk 1]
```diff
@@ -152,7 +152,7 @@
 _BindArguments = Dict[str, Any]
 
 _EntityBindKey = Union[Type[_O], "Mapper[_O]"]
-_SessionBindKey = Union[Type[Any], "Mapper[Any]", "Table"]
+_SessionBindKey = Union[Type[Any], "Mapper[Any]", "Table", str]
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/session.py:155]: I'm going to change this however it requires that I change the code to not use ``__binds`` for this purpose, so I'll do that here.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 93a0cfcb35f83691d2ebf4bafa2b212c74718b64 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 93a0cfcb35f83691d2ebf4bafa2b212c74718b64: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4279

**Comment by zzzeek** [comment]
if this passes, this is good enough for me to take in, I just want to change the `__binds` thing, taht's on my end as that collection shouldnt have been used for this purpose.

**Comment by zzzeek** [comment]
has just one error:


```
pep484 run-test: commands[0] | mypy ./lib/sqlalchemy
lib/sqlalchemy/ext/horizontal_shard.py:165: error: Incompatible types in assignment (expression has type "Optional[Callable[[ORMExecuteState], Iterable[Any]]]", variable has type "Callable[[ORMExecuteState], Any]")  [assignment]
Found 1 error in 1 file (checked 245 source files)
```

File: lib/sqlalchemy/ext/horizontal_shard.py [hunk 2]
```diff
@@ -176,9 +233,13 @@ def _choose_shard_and_assign(self, mapper, instance, **kw):
             state.identity_token = shard_id
         return shard_id
 
-    def connection_callable(
-        self, mapper=None, instance=None, shard_id=None, **kwargs
-    ):
+    def connection_callable(  # type: ignore [override]
```
Comment (g-kisenkov) [inline lib/sqlalchemy/ext/horizontal_shard.py:236]: I get `error: Signature of "connection_callable" incompatible with supertype "Session"  [override]` even if I make the signature exactly the same as specified by `_ConnectionCallableProto` in `lib/sqlalchemy/orm/session.py`.

**Comment by g-kisenkov** [comment]
> has just one error:
> 
> ```
> pep484 run-test: commands[0] | mypy ./lib/sqlalchemy
> lib/sqlalchemy/ext/horizontal_shard.py:165: error: Incompatible types in assignment (expression has type "Optional[Callable[[ORMExecuteState], Iterable[Any]]]", variable has type "Callable[[ORMExecuteState], Any]")  [assignment]
> Found 1 error in 1 file (checked 245 source files)
> ```

Thank you! My VSCode-mypy integration didn't catch it for some reason. I'll be using the `pep484` suite next time.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e40e768492685aa9ce57c4762c571f935e3fd3c7 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset e40e768492685aa9ce57c4762c571f935e3fd3c7 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4279

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

horizontal sharding needs work to be modernized still (or even made moderately useful), but the typing looks great!  thanks for your efforts!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4279

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4279 has been **merged**. Congratulations! :)

**Pull request closed without merging**